wgsl/index.bs

```diff
@@ -6981,13 +6979,296 @@ See [[#statements]] and [[#function-calls]].
+</div>
+
+In more detail, here is what we compute for each function:
+
+* A tag about the control-flow in which the function can be called: one of {AlwaysMustBeUniform, MustBeUniformForControlFlowAfter}
+* A tag for each parameter of the function: one of {AlwaysMustBeUniform, MustBeUniformForControlFlowAfter, MustBeUniformForReturnValue, NoRestriction}
+* A tag for the function as a whole: one of {ControlFlowCannotBeUniformAfter, ReturnValueCannotBeUniform, NoRestriction}
+
+Here is the algorithm for computing this data for a given function:
+
+* Create nodes called "MustBeUniform", "CannotBeUniform", "CF_start", "CF_return", and if the function is non-void a node called "Value_return"
+* Create one node for each parameter of the function which we'll call "arg_i"
+* Walk over the syntax of the function, adding nodes and edges to the graph following the rules of the next section.
+* Look at which nodes are reachable from "MustBeUniform"
+    * If this set includes the node "CannotBeUniform", then reject the program.
+    * If this set includes "CF_start", then the control-flow in which the function can be called is AlwaysMustBeUniform.
+    * Otherwise it is MustBeUniformForControlFlowAfter
+    * For each "arg_i" in this set, the corresponding parameter is AlwaysMustBeUniform
+    * Remove from the graph all nodes that have been visited
+* Look at which nodes are reachable from "CF_return"
+    * If this set includes "CannotBeUniform", then the function is ControlFlowCannotBeUniformAfter
+    * For each "arg_i" in this set, the corresponding parameter is MustBeUniformForControlFlowAfter
+    * Remove from the graph all nodes that have been visited
+* If "Value_return" exists, look at which nodes are reachable from it
+    * If this set includes "CannotBeUniform", then the function is ReturnValueCannotBeUniform
+    * For each "arg_i" in this set, the corresponding parameter is MustBeUniformForReturnValue
+* If the function has not been assigned a tag, then it is NoRestriction.
+* For each parameter, if it has not been assigned a tag, then it is NoRestriction.
+
+Note: The entire graph can be destroyed at this point. The tags listed above are all that we need to remember to analyze callers of this function.
+
+### Uniformity rules for statements ### {#uniformity-statements}
+
+In order to properly deal with reconvergence we must know when there is a single way for control to flow out of a statement.
+This is achieved by following the behaviors analysis (see [[#behaviors]]).
+
+If the set of behaviors for a statement has a single element, this means that we either did not diverge within the statement, or we reconverged, and we pick the node corresponding to control flow at the start of the statement as the node corresponding to control flow at the exit of the statement, overriding the other rules below.
```

Comment:
This doesn't seem quite right. You still need the rules for the sub-analyses, but just avoid applying the final result. For example:
```
fn foo(x : i32) {
  if (x < 5) {
    textureSample(...);
  }
}
```

The use of `textureSample` requires that both the function be called from uniform control flow and that `x` be uniform, but this is not obvious (IMO) from this sentence. It would be good to clarify.

Reply:
Right. I think this whole section is about two things: 
1. The labeling of all the internal nodes.  This needs to be done in any case so the invalidity of the whole function can be checked (MustBeUniform can reach the textureSample call0.
2.  how to generate a control-flow-effect tag for executing a statement, i.e. a post-execution CF.   

This paragraph is about taking the shortcut to compute 2, when the behaviour  analysis for the statement results in a singleton.

Reply:
Secondly, I think the singleton rule is not right here.
The singleton cases for {Break}, {Continue}, {Return}, {Fallthrough} {Discard} should probably not be handled the same as {Next} ?     

The {Next} case seems to be the only intended case.  But I'm not sure.

Reply:
I did not notice how that sentence can be ambiguous. What I meant by "overriding the rules below" is regarding the computation of the tag of the control-flow-after (point 2 in David's comment), nothing else.
I am not sure how to clarify it though?

I intended for the singleton case to also apply to {Break}, etc.., but I don't remember if I had any example where that is useful. I would be willing to make it only apply to {Next}, and suspect it probably would not change anything.